### PR TITLE
Add support for external bundle validators

### PR DESCRIPTION
**Pull request by jmrodri**
**Description of the change:**
Adding support for external bundle validators.

**Motivation for the change:**
Today whenever there are validation changes specific to OpenShift, that often trickles to the upstream release. By adding external bundle validator support, this can allow bundle validator authors to create external binaries to be called by operator-sdk  and allow business specific validators to not impact upstream releases.

**Checklist**

If the pull request includes user-facing changes, extra documentation is required:
- [X] Add a new changelog fragment in `changelog/fragments` (see [`changelog/fragments/00-template.yaml`](https://github.com/operator-framework/operator-sdk/tree/master/changelog/fragments/00-template.yaml))
- [ ] Add or update relevant sections of the docs website in [`website/content/en/docs`](https://github.com/operator-framework/operator-sdk/tree/master/website/content/en/docs)

### CODE REVIEW

File: pkg/validate/external.go
Comment (ryantking) [inline pkg/validate/external.go:32]: This is unnecessary, you can redirect the actual `os.Stderr` variable in the test:

```go
var (
    origStderr io.Writer
    stderrBuf bytes.Buffer
)

BeforeEach(func() {
    origStderr = os.Stderr
    os.Stderr = stderrBuf
})

AfterEach(func() {
    stderrBuf.Reset()
    os.Stderr = origStderr
})
```

Works for all tests both inside and outside the package. Also doesn't require modification of the application code for testing.

File: pkg/validate/external.go
Comment (ryantking) [inline pkg/validate/external.go:69]: More of a nit, but you can avoid reading the entire output of the validator command into memory by using a pipe and asynchronously running the command.

```suggestion
		stdout, err := cmd.StdoutPipe()
		if err != nil {
			return nil, err
		}
		if err := cmd.Start(); err != nil {
		    return nil, err
		}

		// Ensure output conforms to the Output spec.
		dec := json.NewDecoder(stdout)
		dec.DisallowUnknownFields()
		if err := dec.Decode(&manifestresults[i]); err != nil {
			fmt.Printf("decode failed: %v\n", err)
			return nil, err
		}
		if err := cmd.Wait(); err != nil {
		    return nil, err
		}
```

File: pkg/validate/result.go
Comment (ryantking) [inline pkg/validate/result.go:163]: Is there a reason we hard exit here versus returning an error?

**Review by ryantking — COMMENTED** [review]
Looks good, just a few questions/suggestions

File: internal/cmd/operator-sdk/bundle/validate/cmd.go
Comment (jmrodri) [inline internal/cmd/operator-sdk/bundle/validate/cmd.go:102]: `operator-sdk bundle validate ./bundle --alpha-select-external /path/to/external-validator[:/path/to/optional-second-validator:http://github.com/data/my-validator]`

File: pkg/validate/result.go
Comment (jmrodri) [reply pkg/validate/result.go:163]: @ryantking The only thing I saw was that it was a way to short circuit when a failed test was found. I tried removing it and the output changed. As you can see below, by returning an error we end up also logging an error in addition to the test failure. We could just make it an informative error.

### WITHOUT `exit(1)`

Return an error instead. The error we returned was "test failed".

```
operator-sdk bundle validate --alpha-select-external pkg/validate/testdata/fails.sh ~/dev/gatekeeper-operator/bundle
ERRO[0000] Error: Value : invalid field Pesce
FATA[0000] test failed
```

Error code: 1

---

### With the `exit(1)` 

In this case we keep it as is with the `exit(1)` short circuit.

```
operator-sdk bundle validate --alpha-select-external pkg/validate/testdata/fails.sh ~/dev/gatekeeper-operator/bundle
ERRO[0000] Error: Value : invalid field Pesce
```

Error code: 1

Comment (jmrodri) [reply pkg/validate/result.go:163]: I do agree that doing a hard exit seems weird and bad if this gets used some place else as a generic function call.

Comment (ryantking) [reply pkg/validate/result.go:163]: @jmrodri It looks like the root of the problem is that the validator is failing, but not returning an error. I would think `err == nil && !r.Passed` would be impossible. Instead of logging the invalid field, wouldn't we want to bubble it back up and let the caller determine what to do with it?

Comment (jmrodri) [reply pkg/validate/result.go]: @ryantking I changed the method return signature. It now returns `bool, error`. That way there is no adverse `exit`. And the caller can use the boolean value to determine if there were any test failures.

File: pkg/validate/external.go
Comment (jmrodri) [reply pkg/validate/external.go:31]: I left this as is for now.

File: internal/cmd/operator-sdk/bundle/validate/validate.go
Comment (everettraven) [inline internal/cmd/operator-sdk/bundle/validate/validate.go:178]: Nit: Is this `TODO` still something you think should be done?

It might be nice to have this be a function since it has been done before with the `runOptionalValidators` function.

**Review by everettraven — COMMENTED** [review]
Just one nit. Everything else looks good to me.

I don't think it should block it from merging if you don't think the change is necessary so I am not requesting changes. Since I am still learning the ropes I will leave explicit approval to someone else.

File: internal/cmd/operator-sdk/bundle/validate/validate.go
Comment (jmrodri) [reply internal/cmd/operator-sdk/bundle/validate/validate.go:178]: @everettraven for this PR, I wouldn't make this change yet. I am anticipating we might be refactoring things in the future. There is a `RunExternalValidators` that is getting called.

**Review by everettraven — APPROVED** [review]
Everything looks good to me. I am approving and will leave it up to you whether or not you want to wait for another approval before merging.

**Pull request merged**
